hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+        if origin is not None:
+            data['origin'] = origin
+        payload = json.dumps(data)
+        self.send_message(ProtocolMessages.GET_DATA, payload)
+
+    def handle_get_data(self, payload: str) -> None:
+        """Handle a received GET-DATA message."""
+        data = json.loads(payload)
+        txid_hex = data['txid']
+        origin = data.get('origin', '')
+        # self.log.debug('handle_get_data', payload=hash_hex)
+        try:
+            tx = self.protocol.node.tx_storage.get_transaction(bytes.fromhex(txid_hex))
+            self.send_data(tx, origin=origin)
+        except TransactionDoesNotExist:
+            # In case the tx does not exist we send a NOT-FOUND message
+            self.send_message(ProtocolMessages.NOT_FOUND, txid_hex)
+
+    def handle_data(self, payload: str) -> None:
+        """ Handle a received DATA message.
+        """
+        if not payload:
+            return
+        part1, _, part2 = payload.partition(' ')
+        if not part2:
+            origin = None
+            data = base64.b64decode(part1)
+        else:
+            origin = part1
+            data = base64.b64decode(part2)
+
+        try:
+            tx = tx_or_block_from_bytes(data)
+        except struct.error:
+            # Invalid data for tx decode
+            return
+
+        if origin:
+            if origin != 'mempool':
+                # XXX: ban peer?
+                self.protocol.send_error_and_close_connection(f'DATA {origin}: unsupported origin')
+                return
+            assert tx is not None
+            self._on_get_data(tx, origin)
+            return
+
+        assert tx is not None
+        assert tx.hash is not None
+        if self.protocol.node.tx_storage.get_genesis(tx.hash):
+            # We just got the data of a genesis tx/block. What should we do?
+            # Will it reduce peer reputation score?
+            return
+
+        tx.storage = self.protocol.node.tx_storage
+        assert tx.hash is not None
+
+        if self.partial_vertex_exists(tx.hash):
+            # transaction already added to the storage, ignore it
+            # XXX: maybe we could add a hash blacklist and punish peers propagating known bad txs
+            self.manager.tx_storage.compare_bytes_with_local_tx(tx)
+            return
+        else:
+            # If we have not requested the data, it is a new transaction being propagated
+            # in the network, thus, we propagate it as well.
+            if tx.can_validate_full():
+                self.log.info('tx received in real time from peer', tx=tx.hash_hex, peer=self.protocol.get_peer_id())
+                self.on_new_tx(tx, conn=self.protocol, propagate_to_peers=True)
+            else:
+                self.log.info('skipping tx received in real time from peer',
+                              tx=tx.hash_hex, peer=self.protocol.get_peer_id())
+
+    def on_new_tx(self, tx: BaseTransaction, *, conn: Optional['HathorProtocol'] = None,
```

Comment:
It seems some of these arguments are not used inside the method. Is this correct?

Reply:
The signature was moved from `HathorManager.on_new_tx` (not the current one, but a previous version that made these changes there.

I'll clean it up and remove the unused arguments.

Reply:
Done!